### PR TITLE
External information sources

### PR DESCRIPTION
**Pull request by jhoetter**

### CODE REVIEW

**Comment by jhoetter** [comment]
Run this together with:
- [UI PR](https://github.com/code-kern-ai/refinery-ui/pull/51)
- [Proxy PR](https://github.com/code-kern-ai/refinery-gateway-proxy/pull/6)
- [Submodule PR](https://github.com/code-kern-ai/refinery-submodule-model/pull/7)

You can add a model callback using the Python SDK in branch `external-information-sources` ([PR](https://github.com/code-kern-ai/refinery-python-sdk/pull/22)) as follows:
```python
from refinery import Client

user_name = "your-mail" # this is the email you log in with
password = "your-pw"
project_id = "your-clickbait-sample-project-id" # can be found in the URL of the web application

client = Client(user_name, password, project_id, uri="http://localhost:4455")

from sklearn.linear_model import LogisticRegression
clf = LogisticRegression()

from refinery.adapter.sklearn import build_classification_dataset
data = build_classification_dataset(client, "headline", "__clickbait", "distilbert-base-uncased", num_train=50)
clf.fit(data["train"]["inputs"], data["train"]["labels"])

from refinery.callbacks.sklearn import SklearnCallback
callback = SklearnCallback(
    client, 
    clf,
    "clickbait", 
)

callback.run(data["train"]["inputs"], data["train"]["index"])
callback.run(data["test"]["inputs"], data["test"]["index"])
```

File: api/transfer.py [hunk 1]
```diff
@@ -98,6 +99,29 @@ async def post(self, request) -> JSONResponse:
...
+    async def post(self, request) -> JSONResponse:
+        project_id = request.path_params["project_id"]
+        request_body = await request.json()
+        user_id = request_body["user_id"]
+        auth_manager.check_project_access_from_user_id(user_id, project_id)
+        associations = request_body["associations"]
+        indices = request_body["indices"]
+        name = request_body["name"]
+        label_task_name = request_body["label_task_name"]
+        source_type = request_body["source_type"]
```
Comment (JWittmeyer) [inline api/transfer.py]: personal preference: why are these stored in variables when the dict key is = to the variable name & only used once for a function call?

File: controller/information_source/manager.py [hunk 2]
```diff
@@ -8,12 +8,16 @@ def get_information_source(project_id: str, source_id: str) -> InformationSource
...
+    return information_source.get_by_name(project_id, name)
+
+
 def get_all_information_sources(project_id: str) -> List[InformationSource]:
     return information_source.get_all(project_id)
 
 
-def get_overview_data(project_id: str) -> str:
-    return information_source.get_overview_data(project_id)
+def get_overview_data(project_id: str, operator: Optional[str] = "!=") -> str:
```
Comment (JWittmeyer) [inline controller/information_source/manager.py]: optional is only for = None 
[notion link](https://www.notion.so/kern-ai/Coding-Guide-Python-f54c6201aa22471caaff38d765c0d64e#447b4b18e8b94725b8ad0c767445fcb3)

Comment (JWittmeyer) [reply controller/information_source/manager.py]: also, im not the biggest fan of operator as variable since it's pretty undescriptive what it's for.

What do you think about `,for_model_callback: bool = false)->str` and in the information source function something like
``` python
type_selection = " != 'MODEL_CALLBACK'"
if for_model_callback:
     type_selection = " = 'MODEL_CALLBACK'"
```

same for update_is_selected_for_project

File: controller/transfer/association_transfer_manager.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        elif source_type == "heuristic":
+            information_source = information_source_manager.create_information_source(
+                project_id,
+                user_id,
+                labeling_task.id,
+                model_name,
+                "",
+                "This is a heuristic",
+                enums.LabelSource.INFORMATION_SOURCE.value,
+            )
```
Comment (JWittmeyer) [inline controller/transfer/association_transfer_manager.py]: pretty similar maybe:
```suggestion
        
        if source_type == "model_callback":
            description = "This is a model callback"
            type = enums.LabelSource.MODEL_CALLBACK.value
        elif source_type == "heuristic":
            description = "This is a heuristic"
            type = enums.LabelSource.INFORMATION_SOURCE.value
            
        information_source = information_source_manager.create_information_source(
                project_id,
                user_id,
                labeling_task.id,
                model_name,
                "",
                description,
                type,
            )
```

File: controller/transfer/association_transfer_manager.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    )
+    record_ids = [record.id for record in list(records.values())]
+
+    record_label_associations = []
+    for record_id, association in zip(record_ids, associations):
+        label_name, confidence = association
+
+        label = labeling_task_label.get_by_name(
+            project_id, labeling_task.id, label_name
+        )
```
Comment (JWittmeyer) [inline controller/transfer/association_transfer_manager.py]: this will run a new query every loop, i'd go with get all, dict comprehension and grabbing the data then.

This would also open the door for list comprehension for the record_label_associations which will struggle with the append on bigger lists.

Also depending on the size it might make sense to batch this a bit sind an insert with > 10k records will run into issues. maybe max 1k records per batch

File: controller/transfer/association_transfer_manager.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    general.commit()
+
+    try:
+        weak_supervision.calculate_stats_after_source_run(
+            project_id, information_source.id, user_id
+        )
+        notification.send_organization_update(
+            project_id, f"information_source_statistics:{information_source.id}"
+        )
+        general.commit()
```
Comment (JWittmeyer) [inline controller/transfer/association_transfer_manager.py]: unnecessary

File: controller/weak_supervision/weak_supervision_service.py [hunk 6]
```diff
@@ -34,7 +36,10 @@ def calculate_quality_after_labeling(project_id: str, labeling_task_id: str, use
     return service_requests.post_call_or_raise(url, data)
 
 
-def calculate_stats_after_source_run(project_id: str, source_id: str, user_id: str) -> Any:
+@debounce(2)
```
Comment (JWittmeyer) [inline controller/weak_supervision/weak_supervision_service.py]: if i'm not mistaken this was intentionally not debounced to have an "on demand" execution -> debounce will start a thread and not wait for the return. Is this wanted?

File: graphql_api/mutation/information_source.py [hunk 7]
```diff
@@ -91,6 +91,23 @@ def mutate(self, info, project_id: str, value: bool):
...
+        value = graphene.Boolean(required=True)
+
+    ok = graphene.Boolean()
+
+    def mutate(self, info, project_id: str, value: bool):
+        auth.check_demo_access(info)
+        auth.check_project_access(info, project_id)
+        manager.set_all_model_callbacks_selected(project_id, value)
+        notification.send_organization_update(
+            project_id, f"information_source_updated:all"
```
Comment (JWittmeyer) [inline graphql_api/mutation/information_source.py]: is this correct? (information_source_update) this will trigger an update on the heuristics page as well

File: controller/transfer/association_transfer_manager.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+                source_id=information_source.id,
+                return_type=enums.InformationSourceReturnType.RETURN.value,
+                confidence=confidence,
+                created_by=user_id,
+            )
+        )
+
+    record_label_association.delete_by_source_id_and_record_ids(
+        project_id, information_source.id, record_ids
+    )
```
Comment (JWittmeyer) [inline controller/transfer/association_transfer_manager.py:93]: is this wanted? only the provided record ids are removed so a newer import with only 10 record ids would still look like it has a lot of records corresponding to it?

File: api/transfer.py (hunk 1 above)
Comment (jhoetter) [reply api/transfer.py]: true, changed that

File: controller/information_source/manager.py (hunk 2 above)
Comment (jhoetter) [reply controller/information_source/manager.py]: makes sense, changed that

File: controller/transfer/association_transfer_manager.py (hunk 4 above)
Comment (jhoetter) [reply controller/transfer/association_transfer_manager.py]: the batch is limited to 32 files per request. I added the change re the label query calling every loop iteration into one before the loop :)

File: controller/transfer/association_transfer_manager.py (hunk 5 above)
Comment (jhoetter) [reply controller/transfer/association_transfer_manager.py]: removed it

File: controller/weak_supervision/weak_supervision_service.py (hunk 6 above)
Comment (jhoetter) [reply controller/weak_supervision/weak_supervision_service.py]: The data is sent from the SDK in smaller batches, so I don't want to recompute the stats every time (this would again result in an overload). 2 seconds for debounce should be ok I guess. An alternative would be to add another function calling this, which receives the debounce, and only having the callback execute this very function.

Comment (jhoetter) [reply controller/weak_supervision/weak_supervision_service.py]: nvm I changed it so that only the callbacks have a debounce

File: graphql_api/mutation/information_source.py (hunk 7 above)
Comment (jhoetter) [reply graphql_api/mutation/information_source.py]: No, only meant as an update for the model callbacks. i commented it out

File: controller/transfer/association_transfer_manager.py (hunk 8 above)
Comment (jhoetter) [reply controller/transfer/association_transfer_manager.py:93]: again, this is a result of the batching mechanism

**Review by JWittmeyer — APPROVED** [review]
